src/LinuxTracing/LeafFunctionCallManager.cpp

```diff
@@ -0,0 +1,112 @@
+// as the leaf function does not contain frame pointers, the region from $rsp to $rbp will contain
+// the complete innermost frame, plus the locals of the caller's frame (but not the return address).
+// Therefore, libunwindstack will be able to compute two frames, where the outer one is the missing
+// caller and needs to be patched in.
+// As libunwindstack will try to unwind even further, unwinding errors will always be reported.
+// We need to fallback to plausibility checks to detect actual unwinding errors (such as are the
+// frames executable).
+// Note, that we can not simply set libunwindstack to unwind always two frames and compare the outer
+// frame with the respective one in the perf_event_open event, as in case of uprobes overriding the
+// return addresses, both addresses would be identical even if the actual addresses (after uprobe
+// patching) are not.
+// More (internal) documentation on this in: go/stadia-orbit-leaf-frame-pointer
+bool LeafFunctionCallManager::PatchLeafFunctionCaller(CallchainSamplePerfEvent* event,
+                                                      LibunwindstackMaps* current_maps,
+                                                      LibunwindstackUnwinder* unwinder) {
+  CHECK(event != nullptr);
+  CHECK(current_maps != nullptr);
+  CHECK(unwinder != nullptr);
+  CHECK(event->GetCallchainSize() > 2);
+
+  uint64_t stack_size =
+      event->GetRegisters()[PERF_REG_X86_BP] - event->GetRegisters()[PERF_REG_X86_SP];
+  if (stack_size > SAMPLE_STACK_USER_SIZE_128BYTES) {
+    ERROR("Discarding sample where the stack dump is to small (required %lu bytes, got %lu bytes)",
+          stack_size, SAMPLE_STACK_USER_SIZE_128BYTES);
+    return false;
+  }
+
+  const LibunwindstackResult& libunwindstack_result =
+      unwinder->Unwind(event->GetPid(), current_maps->Get(), event->GetRegisters(),
+                       event->GetStackData(), stack_size, true);
+  const std::vector<unwindstack::FrameData>& libunwindstack_callstack =
+      libunwindstack_result.frames();
+
+  if (libunwindstack_callstack.empty()) {
+    ERROR(
+        "Discarding sample as DWARF-based unwinding resulted in empty callstack (error code: %s).",
+        LibunwindstackUnwinder::LibunwindstackErrorString(libunwindstack_result.error_code()));
+    return false;
+  }
+
+  // In case of an intact frame pointer, the region from $rsp to $rbp will only include the locals
+  // of the current frame, and NOT the return address. Thus, unwinding will only report one frame
+  // (the instruction pointer) and we know that the original callchain is already correct.
+  if (libunwindstack_callstack.size() == 1) {
+    return true;
+  }
+
+  // If unwinding results in more than two frames, $rbp was also not set correctly by the caller,
+  // thus frame pointers are not in all non-leaf functions, and our assumptions do not hold.
+  if (libunwindstack_callstack.size() > 2) {
+    ERROR(
```

Comment:
Couldn't this be quite frequent if it just indicated the absence of frame pointers?